syntax = "proto3";

package discord;

message FrecencyItem {
  uint32 total_uses = 1;
  repeated uint64 recent_uses = 2;
  int32 frecency = 3;
  int32 score = 4;
}

message FrecencyUserSettings {
  message Versions {
    uint32 client_version = 1;
    uint32 server_version = 2;
    uint32 data_version = 3;
  }
  Versions versions = 1;

  message FavoriteGIFs {

    message FavoriteGIF {
      enum GIFType {
        NONE = 0;
        IMAGE = 1;
        VIDEO = 2;
      };
      GIFType format = 1;
      string src = 2;
      uint32 width = 3;
      uint32 height = 4;
      uint32 order = 5;
  }

    map<string, FavoriteGIF> gifs = 1;
    bool hide_tooltip = 2;
  }

  FavoriteGIFs favorite_gifs = 2;



  message FavoriteStickers {
    repeated sfixed64 sticker_ids = 1;
  }
  FavoriteStickers favorite_stickers = 3;

  message StickerFrecency {
    map<sfixed64, FrecencyItem> stickers = 1;
  }
  StickerFrecency sticker_frecency = 4;

  message FavoriteEmojis {
    repeated string emojis = 1;
  }
  FavoriteEmojis favorite_emojis = 5;

  message EmojiFrecency {
    map<string, FrecencyItem> emojis = 1;
  }
  EmojiFrecency emoji_frecency = 6;

  message ApplicationCommandFrecency {
    map<string, FrecencyItem> application_commands = 1;
  }
  ApplicationCommandFrecency application_command_frecency = 7;

}

// I am not sure the variable types are correct
// Nor am I sure how to handle repeat: 2 and etc
// As a reference, I used protobuf-ts SRC and
// https://googleapis.dev/python/protobuf/latest/google/protobuf/descriptor.html#google.protobuf.descriptor.FieldDescriptor.ProtoTypeToCppProtoType
// The proto declaration is painfully 'stolen' from Discord's asset at 72001f09f5dc9cab20d0.js
// At least I now know how to use protobuf and debug webpacker chunked projects, sigh
